File: python/claude/p3_nc.py
DEST_CAPACITY = 256


def copy_string(src: str | None) -> bytearray | None:
    """
    Copies the string src into a fixed-size destination buffer using its
    computed byte length.  Appends a zero-byte null terminator.

    Returns a bytearray on success, None on any validation failure.
    """
    if src is None:
        print("Error: null source argument.")
        return None
    if not isinstance(src, str):
        print("Error: src must be a str.")
        return None

    src_bytes = src.encode("utf-8")
    n = len(src_bytes)

    # n + null terminator must fit within capacity
    if n >= DEST_CAPACITY:
        print(
            f"Error: source byte length ({n}) would overflow destination "
            f"capacity ({DEST_CAPACITY}) after null terminator."
        )
        return None

    dst = bytearray(DEST_CAPACITY)
    dst[:n] = src_bytes
    dst[n] = 0

    print(f'Success: copied "{src}" ({n} byte(s)).')
    return dst


if __name__ == "__main__":
    # Valid call
    result = copy_string("Hello")
    if result is not None:
        print("First 5 bytes copied successfully.")

    # Invalid: null source
    copy_string(None)

File: java/claude/p3_nc.java
import java.nio.charset.StandardCharsets;

public class p3_nc {

    private static final int DEST_CAPACITY = 256;

    /**
     * Copies the string {@code src} into a fixed-size destination buffer
     * using its computed byte length.  Appends a zero-byte null terminator.
     *
     * @param src source string (must be non-null)
     * @return    the destination byte array on success, null on any validation failure
     */
    public static byte[] copyString(String src) {
        if (src == null) {
            System.err.println("Error: null source argument.");
            return null;
        }

        byte[] srcBytes = src.getBytes(StandardCharsets.UTF_8);
        int n = srcBytes.length;

        /* n + null terminator must fit within capacity */
        if (n >= DEST_CAPACITY) {
            System.err.println("Error: source byte length (" + n
                    + ") would overflow destination capacity (" + DEST_CAPACITY
                    + ") after null terminator.");
            return null;
        }

        byte[] dst = new byte[DEST_CAPACITY];
        System.arraycopy(srcBytes, 0, dst, 0, n);
        dst[n] = 0;

        System.out.println("Success: copied \"" + src + "\" (" + n + " byte(s)).");
        return dst;
    }

    public static void main(String[] args) {
        /* Valid call */
        byte[] result = copyString("Hello");
        if (result != null) {
            System.out.println("First 5 bytes copied successfully.");
        }

        /* Invalid: null source */
        copyString(null);
    }
}
